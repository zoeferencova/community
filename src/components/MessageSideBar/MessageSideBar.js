import React, { Component } from "react";
import CommUnityContext from "../../contexts/context";
import { ProfilePicture } from "../../components/Utils/Utils";
import styles from "../MessageLayout/MessageLayout.module.css";

export default class MessageSideBar extends Component {
    static contextType = CommUnityContext;

    makeChats = () => {
        const { chats, activeChat, setActiveChat } = this.props;
        const sortedChats = chats.sort((a, b) => {
            if (b.messages === undefined || a.messages === undefined) {
                return -1
            } else if (b.messages[b.messages.length - 1].message_timestamp < a.messages[a.messages.length - 1].message_timestamp) {
                return -1
            } else if (b.messages[b.messages.length - 1].message_timestamp > a.messages[a.messages.length - 1].message_timestamp) {
                return 1
            }
            return null;
        })

        return sortedChats.map(chat => {
            const lastMessage = chat.messages ? chat.messages[chat.messages.length - 1] : "";
            
            const user = chat.user1.id === this.context.user.id ? chat.user2 : chat.user1;
            const classNames = (activeChat && activeChat.id === chat.id) ? styles.active : ""
            return (
                <div key={chat.id} className={`${styles.user} ${classNames}`} onClick={() => { setActiveChat(chat) }}>
                    <ProfilePicture className={styles.sidebarPic} first_name={user.first_name} />
                    <div className={styles.userInfo}>
                        <div className={styles.name}>{user.first_name}</div>
                        {lastMessage && <div className={styles.lastMessage}>{lastMessage.message_content}</div>}
                    </div>
                </div>
            )
        })
    }

    render() {
        const { setActiveChat } = this.props;
        return (
            <div id="side-bar" className={`${styles.sideBar} ${this.props.mobileDisplay === true ? styles.activeMobile : styles.inactiveMobile}`}>
                <div className={styles.heading}>
                    <div className={styles.chatName}>Chats</div>
                </div>
                <div className={styles.users} ref="users" onClick={e => {(e.target === this.refs.user) && setActiveChat(null)}}>
                    {this.makeChats()}
                </div>
            </div>
        )
    }
}