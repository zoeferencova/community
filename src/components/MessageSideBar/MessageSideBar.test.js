import React from "react";
import { render } from "enzyme";
import MessageSideBar from "./MessageSideBar";
import CommUnityContext from "../../contexts/context";
import mockContext from "../../contexts/mock-context";

describe("MessageSideBar", () => {
    it("renders MessageSideBar component given mock context data", () => {     
        const mockProps = {
            chats: mockContext.chats,
            activeChat: mockContext.activeChat,
            setActiveChat: jest.fn()
        }
        
        const wrapper = render(
            <CommUnityContext.Provider value={mockContext}>
                <MessageSideBar {...mockProps} />
            </CommUnityContext.Provider>
        ); 

        expect(wrapper).toMatchSnapshot();
    });
});