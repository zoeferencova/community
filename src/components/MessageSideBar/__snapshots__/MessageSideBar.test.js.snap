// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageSideBar renders MessageSideBar component given mock context data 1`] = `
<MessageSideBar
  activeChat={
    Object {
      "id": 25,
      "messages": Array [
        Object {
          "chat_id": 25,
          "id": 285,
          "message_content": "hi",
          "message_timestamp": "2020-10-06T10:46:45-04:00",
          "sender_id": 1,
        },
        Object {
          "chat_id": 25,
          "id": 286,
          "message_content": "hey",
          "message_timestamp": "2020-10-06T10:46:45-04:00",
          "sender_id": 2,
        },
      ],
      "post": Object {
        "date_created": "2020-10-06T10:46:45-04:00",
        "description": null,
        "id": 4,
        "post_type": "offer",
        "urgency": null,
        "user_id": 2,
      },
      "user1": Object {
        "first_name": "Zoe",
        "id": 1,
      },
      "user2": Object {
        "first_name": "James",
        "id": 2,
      },
    }
  }
  chats={
    Array [
      Object {
        "id": 25,
        "messages": Array [
          Object {
            "chat_id": 25,
            "id": 285,
            "message_content": "hi",
            "message_timestamp": "2020-10-06T10:46:45-04:00",
            "sender_id": 1,
          },
          Object {
            "chat_id": 25,
            "id": 286,
            "message_content": "hey",
            "message_timestamp": "2020-10-06T10:46:45-04:00",
            "sender_id": 2,
          },
        ],
        "post": Object {
          "date_created": "2020-10-06T10:46:45-04:00",
          "description": null,
          "id": 4,
          "post_type": "offer",
          "urgency": null,
          "user_id": 2,
        },
        "user1": Object {
          "first_name": "Zoe",
          "id": 1,
        },
        "user2": Object {
          "first_name": "James",
          "id": 2,
        },
      },
    ]
  }
  setActiveChat={[MockFunction]}
>
  <div
    className="sideBar inactiveMobile"
    id="side-bar"
  >
    <div
      className="heading"
    >
      <div
        className="chatName"
      >
        Chats
      </div>
    </div>
    <div
      className="users"
      onClick={[Function]}
    >
      <div
        className="user active"
        id="chat25"
        key="25"
        onClick={[Function]}
      >
        <ProfilePicture
          className="sidebarPic"
          first_name="James"
        >
          <div
            className="circle indigo sidebarPic"
            first_name="James"
          >
            J
          </div>
        </ProfilePicture>
        <div
          className="userInfo"
        >
          <div
            className="name"
          >
            James
          </div>
          <div
            className="lastMessage"
          >
            hey
          </div>
        </div>
      </div>
    </div>
  </div>
</MessageSideBar>
`;
